### PR TITLE
Project Recovery - Background Saving

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
This PR implements the saving functionality for background saving. The following changes are included:

- Added a function to the Python API to write workspace histories
- Updated `ProjectSerialiser` to skip attempting file loads in recovery mode
- Refactored `ProjectSeraliser` to remove any workspaces not in the project file, in recovery mode
- Added `ProjectRecoveryThread` which contains a background thread that saves out all workspace histories and GUI windows periodically
- These are all saved to `%appdata%/recovery/*timestamp*` a maximum of 5 are kept in rotation
- Saves every minute (configurable)
- `ProjectRecoveryThread` will stop if anything fails to work for any reason.
- Added new configuration keys for all of these options
- Information logging indicates when saving happens

*To Do in a future PR*
- Recovery steps (loading)
- Folder clean-up on clean Mantid exit
- Resolve some window listeners not being correctly linked, so clear all memory leaves GUI windows behind
- Handle `secondsBetween <= 0`

**To test:**
- *Optional* set `projectRecovery.secondsBetween = 60` in the properties file to a lower value (bottom of file)
- Open `%appdata%` folder on your machine
- Open Mantid Plot
- Set debugging to ~notice~ debug
- Check that saving triggers correctly
- Ensure after 5 saves the folders in `recovery` are being rotated
- Create a sample workspace and open a GUI
- Check the saved files in the timestamped folder (project file and .py) contain the new GUI / WS

-----------------------------

- Add a std::exception anywhere in the main thread body
- Check it is caught correctly and the thread stops saving
- Add a std::exception anywhere in the ProjectSerialiser save body
- *same again*
- Add an error to the Python simple API changes (write_all_workspaces_histories)
- *same again*
- Check the configuration key stops project saving when `projectRecovery.enabled` is set to false

-----------------------------

- Open a large number of GUI windows
- Open a large number of workspaces (GenerateMess can be used)
- Check that the save triggering does not cause any freezes or hangs

Does this update require release notes?
- [ ] Yes
- [x] No

Release notes will be updated as part of the loading mechanism in the future PR

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidPlot/src/ProjectRecoveryThread.cpp [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+const int NO_OF_CHECKPOINTS =
+    getConfigValue<int>(NO_OF_CHECKPOINTS_KEY).get_value_or(5);
+
+// Implementation variables
+Mantid::Kernel::Logger g_log("Project Recovery Thread");
+const std::chrono::seconds TIME_BETWEEN_SAVING(SAVING_TIME);
+
+} // namespace
+
+namespace MantidQt {
```
Comment (martyngigg) [inline MantidPlot/src/ProjectRecoveryThread.cpp]: Usually namespaces are good practice but as this is currently only part of MantidPlot so can we remove the namespaces as our standard has been that `MantidQt::API` would be part of the `MantidQtAPI` library.

File: MantidPlot/src/ProjectRecoveryThread.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+File change history is stored at: <https://github.com/mantidproject/mantid>
+*/
+
+namespace MantidQt {
+namespace API {
+class ProjectRecoveryThread {
```
Comment (martyngigg) [inline MantidPlot/src/ProjectRecoveryThread.h]: Being picky I would probably just go with `ProjectRecovery` as the class name as I think threading is an implementation detail.

File: scripts/Recovery/GenerateHistory.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import sys
+from mantid.simpleapi import *
+
+sys.path.append('/Applications/MantidPlot_312.app/Contents/MacOS')
+
+Load(Filename='/Users/keithbutler/Documents/Mantid/Data/Muons/Diffusion/EMU00011892.nxs',
```
Comment (Mantid-Matthew) [inline scripts/Recovery/GenerateHistory.py]: Should this be committed filepaths look specific to one user?

File: scripts/Recovery/GenerateMess.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+from __future__ import print_function
+import sys
+from mantid.simpleapi import *
+
+sys.path.append('/Applications/MantidPlot_312.app/Contents/MacOS')
+
+data_directory = '/Users/keithbutler/Documents/Mantid/Data/Muons/Diffusion/EMU000'
```
Comment (Mantid-Matthew) [inline scripts/Recovery/GenerateMess.py]: Should this file be committed?

File: MantidPlot/src/ProjectRecoveryThread.cpp [hunk 5]
```diff
@@ -0,0 +1,326 @@
+#include "ProjectRecoveryThread.h"
+
+#include "ApplicationWindow.h"
+#include "Folder.h"
+#include "ProjectSerialiser.h"
+#include "globals.h"
```
Comment (martyngigg) [inline MantidPlot/src/ProjectRecoveryThread.cpp]: Do we need anything our of the globals header?

File: scripts/Recovery/OrderedHistory.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
+from matid.simpleapi import *
+
+Load(Filename='/Users/keithbutler/Documents/Mantid/Data/Muons/Diffusion/EMU00011892.nxs', OutputWorkspace='EMU00011892',
```
Comment (Mantid-Matthew) [inline scripts/Recovery/OrderedHistory.py]: Should these paths be less specific?

File: Framework/PythonInterface/mantid/simpleapi.py [hunk 7]
```diff
@@ -1407,3 +1407,49 @@ def _attach_algorithm_func_as_method(method_name, algorithm_wrapper, algm_object
                                              algm_object.workspaceMethodOn())
 
 # -------------------------------------------------------------------------------------------------------------
+
+
+def write_all_workspaces_histories(folder):
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/simpleapi.py]: Can you help me understand why we wrote the code in Python and not in C++? At this level these functions will be available to all users automatically - was that the intention?

I can't quite see why we have some code here and then code in the `scripts/Recovery` directory.

File: scripts/Recovery/GenerateHistory.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import sys
```
Comment (martyngigg) [inline scripts/Recovery/GenerateHistory.py]: Was this file meant to be added as it looks a bit like some example code or a test?

File: scripts/Recovery/GenerateMess.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
+from __future__ import print_function
```
Comment (martyngigg) [inline scripts/Recovery/GenerateMess.py]: Was this file meant to be added as it looks a bit like some example code or a test?

File: scripts/Recovery/OrderedHistory.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
+from matid.simpleapi import *
```
Comment (martyngigg) [inline scripts/Recovery/OrderedHistory.py]: Was this file meant to be added as it looks a bit like some example code or a test?

File: scripts/Recovery/OrderWorkspaceHistory.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    with open(infile) as f:
+        reader = csv.reader(f, delimiter='#')
+        commands.append([tuple(r) for r in reader])
+
+# Add lists of histories together
+all_commands = [i for sublist in commands for i in sublist]
+
+# Remove duplicate commands
+if _has_duplicates(all_commands):
+    all_unique_commands = _remove_duplicates(all_commands)
```
Comment (Mantid-Matthew) [inline scripts/Recovery/OrderWorkspaceHistory.py]: Why would duplicate commands ever occur ?

File: scripts/Recovery/SampleRecoverySave.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
+from mantid.simpleapi import *
```
Comment (martyngigg) [inline scripts/Recovery/SampleRecoverySave.py]: Was this file meant to be added as it looks a bit like some example code or a test? For samples can we add an `examples` directory to split of functional code and example scripts?

**Review by martyngigg — COMMENTED** [review]
A few high level comments before I go through the recovery code in more depth.

Generally this looks really good - I just have a few questions regarding code structure.

**Review by Matthew-Andrew — COMMENTED** [review]
Looks good in general a few comments which are mainly questions I had. Also are there any automated tests which would be appropriate for these changes?

File: MantidPlot/src/ProjectRecoveryThread.cpp (hunk 5 above)
Comment (DavidFair) [reply MantidPlot/src/ProjectRecoveryThread.cpp]: Not in this PR, but in the next one (for project recovery) the load method in `ProjectSerialiser` requires the mantid version as a parameter

File: Framework/PythonInterface/mantid/simpleapi.py (hunk 7 above)
Comment (DavidFair) [reply Framework/PythonInterface/mantid/simpleapi.py]: Originally this was based on spike work written in Python

As we now only save out WS histories with the timestamp concatenated I agree we move this element to C++.

In a future PR I've converted `OrderWorkspaceHistories` into a utility algorithm which we can execute directly too.

File: MantidPlot/src/ProjectRecoveryThread.cpp (hunk 5 above)
Comment (martyngigg) [reply MantidPlot/src/ProjectRecoveryThread.cpp]: I see. In that case we should use the `MantidKernel/MantidVersion.h` header which is the proper mantid versioning information. That file is from the original `qtiplot` and so it won't quite give the numbers you would expect.

File: Framework/PythonInterface/mantid/simpleapi.py (hunk 7 above)
Comment (martyngigg) [reply Framework/PythonInterface/mantid/simpleapi.py]: :+1:

File: MantidPlot/src/ProjectSerialiser.cpp [hunk 13]
```diff
@@ -478,8 +587,12 @@ QString ProjectSerialiser::saveWorkspaces() {
       wsNames += "\t";
       wsNames += wsName;
 
-      std::string fileName(workingDir + "//" + wsName.toStdString() + ".nxs");
-      window->mantidUI->savedatainNexusFormat(fileName, wsName.toStdString());
+      // Skip saving in project recovery mode
```
[the remark this answers is not in the document]
Comment (DavidFair) [reply MantidPlot/src/ProjectSerialiser.cpp]: Yes, we save out the histories as a Python Script in `ProjectRecovery` instead. so we ask the serialiser to skip saving in this special case

File: MantidPlot/src/ApplicationWindow.cpp [hunk 14]
```diff
@@ -4628,66 +4618,44 @@ ApplicationWindow *ApplicationWindow::openProject(const QString &filename,
...
-  fileTS.setCodec(QTextCodec::codecForName("UTF-8"));
-
-  QString baseName = fileInfo.fileName();
-
-  // Skip mantid version line
-  fileTS.readLine();
-
-  // Skip the <scripting-lang> line. We only really use python now anyway.
-  fileTS.readLine();
-  setScriptingLanguage("Python");
```
[the remark this answers is not in the document]
Comment (DavidFair) [reply MantidPlot/src/ApplicationWindow.cpp]: 👍

File: MantidPlot/src/ProjectRecoveryThread.cpp (hunk 5 above)
Comment (DavidFair) [reply MantidPlot/src/ProjectRecoveryThread.cpp]: Okay I'll pick that up in the loading PR, as this is only relevant there.
I've removed the header in the mean time

Comment (martyngigg) [reply MantidPlot/src/ProjectRecoveryThread.cpp]: :+1:

**Comment by DavidFair** [comment]
@martyngigg @Mantid-Matthew 
This should be ready for additional reviews whilst I nurse it through the build servers / check it works on GCC-4.x

Changes:
- Removed script from initial spikework
- Removed script from simpleapi
- Uses a C++ algorithm to save out workspaces (making future work easier too)
- Renamed `ProjectRecoveryThread` -> `ProjectRecovery`

File: MantidPlot/src/ApplicationWindow.cpp [hunk 15]
```diff
@@ -4628,66 +4619,44 @@ ApplicationWindow *ApplicationWindow::openProject(const QString &filename,
...
-  fileTS.readLine();
-
   folders->blockSignals(true);
   blockSignals(true);
 
+  // Open as a top level folder
+  ProjectSerialiser serialiser(this);
+  try {
+    serialiser.load(filename.toStdString(), fileVersion);
+  } catch (std::runtime_error &e) {
```
Comment (martyngigg) [inline MantidPlot/src/ApplicationWindow.cpp]: Is this guaranteed to only throw a `runtime_error`?

Comment (DavidFair) [reply MantidPlot/src/ApplicationWindow.cpp]: It's the only exception type that we throw from `ProjectSerialiser`. 
Anything else is probably a hard error (such as OOM) we probably don't want to attempt any recovery from

Comment (martyngigg) [reply MantidPlot/src/ApplicationWindow.cpp]: :+1:

File: MantidPlot/src/ProjectRecovery.cpp [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+/// Returns a string to the current top level recovery folder
+std::string getRecoveryFolder() {
+  static std::string recoverFolder =
+      Mantid::Kernel::ConfigService::Instance().getAppDataDir() + "/recovery/";
+  return recoverFolder;
+}
+
+/// Gets a formatted timestamp
+std::string getTimeStamp() {
+  const char *formatSpecifier = "%Y-%m-%d %H-%M-%S";
```
Comment (martyngigg) [inline MantidPlot/src/ProjectRecovery.cpp]: Maybe this is a personal thing but spaces in these paths can cause confusion and be slightly harder to work with in the terminal. Could we use `T` as a separator for than a space as in the ISO 8601 standard?

File: MantidPlot/src/ProjectRecovery.cpp [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+      g_log.debug("Project Recovery: Stopping background saving thread");
+      return;
+    }
+
+    g_log.debug("Project Recovery: Saving started");
+    // "Timeout" - Save out again
+    // Generate output paths
+    const auto basePath = getOutputPath();
+
+    Poco::File(basePath).createDirectory();
```
Comment (martyngigg) [inline MantidPlot/src/ProjectRecovery.cpp]: On my OS, Ubuntu 16.04, I get an exception in the log after the first time project recovery saving kicks in:
```
Project recovery has stopped. Please report this to the development team.
Exception:
File not found
```
I think the issue is that the on linux, by default, when you create a directory the parent directories must exist but in my case the `$HOME/.mantid/recovery` doesn't exist so it cannot create the timestamped child directories.
I tested this by changing this line to `Poco::File(basePath).createDirectories()` and this seems to work.

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by martyngigg — APPROVED** [review]
This looks good now and produces what I would expect given the exceptions mentioned for the future work.

LGTM

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Could someone take a look at merging this soonish? David has some other changes that are to be reviewed and having these in master will help the PR diff.

**Pull request merged**